x/evm/keeper/keeper.go

```diff
@@ -305,3 +307,31 @@ func (k Keeper) AddTransientGasUsed(ctx sdk.Context, gasUsed uint64) (uint64, er
 	k.SetTransientGasUsed(ctx, result)
 	return result, nil
 }
+
+// SetHeaderHash stores the hash of the current block header in the store.
+func (k Keeper) SetHeaderHash(ctx sdk.Context) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixHeaderHash)
+	key := make([]byte, 8)
+	height, err := ethermint.SafeUint64(ctx.BlockHeight())
+	if err != nil {
+		panic(err)
+	}
+	binary.BigEndian.PutUint64(key, height)
+	store.Set(key, ctx.HeaderHash())
+}
+
+// GetHeaderHash retrieves the hash of a block header from the store by height.
+func (k Keeper) GetHeaderHash(ctx sdk.Context, height uint64) []byte {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixHeaderHash)
```

Comment:
return `ctx.HeaderHash()` if it's not empty and `height == ctx.BlockHeight()`, is still a good optimisation?